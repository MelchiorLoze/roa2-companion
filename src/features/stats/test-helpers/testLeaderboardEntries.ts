import { type LeaderboardEntry } from '../types/rank';

export const testLeaderboardEntries: LeaderboardEntry[] = [
  // Aetherean
  { steamId: 1, position: 1, elo: 2162 },
  { steamId: 2, position: 2, elo: 2089 },
  { steamId: 3, position: 3, elo: 1910 },
  { steamId: 4, position: 4, elo: 1837 },
  // Grandmaster
  { steamId: 5, position: 5, elo: 1791 },
  // Master
  { steamId: 6, position: 6, elo: 1655 },
  { steamId: 7, position: 7, elo: 1649 },
  { steamId: 8, position: 8, elo: 1523 },
  // Diamond
  { steamId: 9, position: 9, elo: 1450 },
  { steamId: 10, position: 10, elo: 1300 },
  // Platinum
  { steamId: 11, position: 11, elo: 1258 },
  { steamId: 12, position: 12, elo: 1111 },
  // Gold
  { steamId: 13, position: 13, elo: 1000 },
  { steamId: 14, position: 14, elo: 920 },
  // Silver
  { steamId: 15, position: 15, elo: 820 },
  { steamId: 16, position: 16, elo: 752 },
  // Bronze
  { steamId: 17, position: 17, elo: 699 },
  { steamId: 18, position: 18, elo: 510 },
  // Stone
  { steamId: 19, position: 19, elo: 400 },
  { steamId: 20, position: 20, elo: 300 },
  { steamId: 25, position: 25, elo: -100 },
];
