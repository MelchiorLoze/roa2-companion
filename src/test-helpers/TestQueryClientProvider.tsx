import { QueryClient, type QueryClientConfig, QueryClientProvider } from '@tanstack/react-query';
import { type PropsWithChildren } from 'react';

const queryClientTestConfig: QueryClientConfig = {
  defaultOptions: {
    queries: {
      retry: false,
      gcTime: 0,
      staleTime: 0,
    },
    mutations: {
      retry: false,
      gcTime: 0,
    },
  },
};

export const TestQueryClientProvider = ({ children }: PropsWithChildren) => {
  const queryClient = new QueryClient(queryClientTestConfig);

  return <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>;
};
