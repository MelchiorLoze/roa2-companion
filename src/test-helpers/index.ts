import { Category, type Item, Rarity } from '@/types/item';

export { createItemDto } from './createItemDto';
export { TestQueryClientProvider } from './TestQueryClientProvider';

export const testItemList: Item[] = [
  {
    id: '1',
    name: 'Item 1',
    rarity: Rarity.COMMON,
    category: Category.ICON,
    coinPrice: 2000,
    buckPrice: 20,
  },
  {
    id: '2',
    name: 'Item 2',
    rarity: Rarity.EPIC,
    category: Category.DEATHEFFECT,
    buckPrice: 500,
  },
  {
    id: '3',
    name: 'Item 3',
    rarity: Rarity.RARE,
    category: Category.EMOTE,
    coinPrice: 1500,
  },
  {
    id: '4',
    name: 'Item 4',
    rarity: Rarity.LEGENDARY,
    category: Category.SKIN,
  },
  {
    id: '5',
    name: 'Item 5',
    rarity: Rarity.COMMON,
    category: Category.STAGESKIN,
    coinPrice: 200000,
    buckPrice: 2000,
  },
];
