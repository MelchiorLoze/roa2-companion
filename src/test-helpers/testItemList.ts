import { Category, type Item, Rarity } from '@/types/item';

export const testItemList: Item[] = [
  {
    id: '1',
    friendlyId: 'Icon_item_1',
    name: 'Item 1',
    rarity: Rarity.COMMON,
    category: Category.ICON,
    coinPrice: 2000,
    buckPrice: 20,
  },
  {
    id: '2',
    friendlyId: 'DeathEffect_item_2',
    name: 'Item 2',
    rarity: Rarity.EPIC,
    category: Category.DEATHEFFECT,
    buckPrice: 500,
  },
  {
    id: '3',
    friendlyId: 'Emote_item_3',
    name: 'Item 3',
    rarity: Rarity.RARE,
    category: Category.EMOTE,
    coinPrice: 1500,
  },
  {
    id: '4',
    friendlyId: 'Skin_item_4',
    name: 'Item 4',
    rarity: Rarity.LEGENDARY,
    category: Category.SKIN,
  },
  {
    id: '5',
    friendlyId: 'StageSkin_item_5',
    name: 'Item 5',
    rarity: Rarity.COMMON,
    category: Category.STAGESKIN,
    coinPrice: 200000,
    buckPrice: 2000,
  },
];
