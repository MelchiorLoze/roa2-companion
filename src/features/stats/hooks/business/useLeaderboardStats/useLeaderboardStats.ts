import { type barDataItem, type lineDataItem } from 'react-native-gifted-charts';
import { useUnistyles } from 'react-native-unistyles';
import { type UnistylesTheme } from 'react-native-unistyles/lib/typescript/src/types';

import { useSeason } from '@/features/stats/contexts/SeasonContext/SeasonContext';
import { getRank, type LeaderboardEntry, MAX_AETHEREAN_PLAYERS, type Rank } from '@/types/rank';

import { useCommunityLeaderboard } from '../../data/useCommunityLeaderboard/useCommunityLeaderboard';

const DISTRIBUTION_PRECISION = 10;

const roundedElo = (elo: number) => Math.floor(elo / DISTRIBUTION_PRECISION) * DISTRIBUTION_PRECISION;

type RankDistributionBarDataItem = barDataItem;
const getRankDistribution = (
  leaderboardEntries: LeaderboardEntry[],
  theme: UnistylesTheme,
): RankDistributionBarDataItem[] => {
  const result = leaderboardEntries.reduce(
    (acc, entry) => {
      const rank = getRank(entry.elo, entry.position);
      if (acc[rank]?.value) acc[rank].value++;
      else acc[rank] = { value: 1, frontColor: theme.color[rank] };
      return acc;
    },
    {} as Record<Rank, RankDistributionBarDataItem>,
  );

  return Object.values(result);
};

type EloDistributionLineDataItem = { elo: number } & lineDataItem;
const getEloDistribution = (
  leaderboardEntries: LeaderboardEntry[],
  firstPlayerElo: number,
  lastPlayerElo: number,
  userElo: number,
): EloDistributionLineDataItem[] => {
  const totalEloRange = firstPlayerElo - lastPlayerElo + 1;
  const result = Array.from(
    { length: Math.ceil(totalEloRange / DISTRIBUTION_PRECISION) },
    (_, i) => i * DISTRIBUTION_PRECISION + roundedElo(lastPlayerElo),
  ).reduce(
    (acc, elo) => ({ ...acc, [elo]: { elo: elo, value: 0 } as EloDistributionLineDataItem }),
    {} as Record<number, EloDistributionLineDataItem>,
  );

  leaderboardEntries.forEach((entry) => {
    const elo = roundedElo(entry.elo);
    if (result[elo].value) result[elo].value++;
    else result[elo].value = 1;
  });

  const userEloRounded = roundedElo(userElo);
  if (result[userEloRounded]) result[userEloRounded].dataPointColor = 'red';

  return Object.values(result).sort((a, b) => a.elo - b.elo);
};

export const useLeaderboardStats = (userElo = 0) => {
  const { leaderboardId } = useSeason();
  const { leaderboardEntries, isLoading: isLoadingLeaderboard } = useCommunityLeaderboard(leaderboardId);

  const { theme } = useUnistyles();

  const firstPlayerElo = leaderboardEntries[0]?.elo ?? 0;
  const lastPlayerElo = Math.min(0, leaderboardEntries[leaderboardEntries.length - 1]?.elo ?? 0);
  const lastAethereanElo =
    leaderboardEntries.findLast((entry) => entry.elo > 1800 && entry.position < MAX_AETHEREAN_PLAYERS)?.elo ?? 0;

  return {
    firstPlayerElo,
    lastPlayerElo,
    lastAethereanElo,
    rankDistribution: getRankDistribution(leaderboardEntries, theme),
    eloDistribution: getEloDistribution(leaderboardEntries, firstPlayerElo, lastPlayerElo, userElo),
    isLoading: isLoadingLeaderboard,
  };
};
