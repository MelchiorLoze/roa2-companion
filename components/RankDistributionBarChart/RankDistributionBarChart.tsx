import { Text } from 'react-native';
import { BarChart } from 'react-native-gifted-charts';
import { StyleSheet } from 'react-native-unistyles';

import { Spinner } from '@/components';
import { useLeaderboardStats } from '@/hooks/business/useLeaderboardStats/useLeaderboardStats';
import { Rank, RANK_ELO_INTERVALS } from '@/types/rank';

// Caclulate the widths of the chart bars to be proportional to the elo intervals
const getBarWidths = (firstPlayerElo: number, lastPlayerElo: number, lastAethereanElo: number, totalWidth: number) =>
  [
    RANK_ELO_INTERVALS[Rank.STONE].max - lastPlayerElo + 1, // Assuming there could be players below 0 elo
    RANK_ELO_INTERVALS[Rank.BRONZE].size(),
    RANK_ELO_INTERVALS[Rank.SILVER].size(),
    RANK_ELO_INTERVALS[Rank.GOLD].size(),
    RANK_ELO_INTERVALS[Rank.PLATINUM].size(),
    RANK_ELO_INTERVALS[Rank.DIAMOND].size(),
    RANK_ELO_INTERVALS[Rank.MASTER].size(),
    lastAethereanElo - RANK_ELO_INTERVALS[Rank.GRANDMASTER].min + 1,
    firstPlayerElo - lastAethereanElo,
  ].map((eloIntervalSize) => (eloIntervalSize / firstPlayerElo) * totalWidth);

type Props = {
  width: number;
};

export const RankDistributionBarChart = ({ width }: Props) => {
  const {
    firstPlayerElo,
    lastPlayerElo,
    lastAethereanElo,
    rankDistribution,
    isLoading: isLoadingLeaderboard,
  } = useLeaderboardStats();

  if (isLoadingLeaderboard) return <Spinner />;

  const barWidths = getBarWidths(firstPlayerElo, lastPlayerElo, lastAethereanElo, width);

  return (
    <BarChart
      data={rankDistribution.reverse().map((item, index) => ({
        ...item,
        barWidth: barWidths[index] % 100,
        topLabelComponent: () => <Text style={styles.topLabel}>{item.value}</Text>,
      }))}
      disablePress
      disableScroll
      height={width}
      hideAxesAndRules
      spacing={0}
      width={width}
      xAxisLabelsHeight={0}
      yAxisLabelWidth={0}
    />
  );
};

const styles = StyleSheet.create((theme) => ({
  topLabel: {
    fontFamily: theme.font.secondary.black,
    fontSize: 10,
    color: theme.color.black,
  },
}));
