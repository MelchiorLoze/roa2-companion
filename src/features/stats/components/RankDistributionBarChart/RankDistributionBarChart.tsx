import { Text, View } from 'react-native';
import type { barDataItem } from 'react-native-gifted-charts';
import { BarChart } from 'react-native-gifted-charts';
import { StyleSheet } from 'react-native-unistyles';

import { Spinner } from '@/components/Spinner/Spinner';

import { useLeaderboardStats } from '../../hooks/business/useLeaderboardStats/useLeaderboardStats';
import { Rank, RANK_ELO_INTERVALS } from '../../types/rank';

const formatNumber = (value: number) => {
  if (value >= 1_000_000) return `${(value / 1_000_000).toFixed(1)}M`;
  if (value >= 1_000) return `${(value / 1_000).toFixed(1)}K`;
  return value.toString();
};

// Caclulate the widths of the chart bars to be proportional to the elo intervals
const getBarWidths = (firstPlayerElo: number, lastPlayerElo: number, lastAethereanElo: number, totalWidth: number) => {
  const totalEloRange = firstPlayerElo - lastPlayerElo + 1;
  return [
    RANK_ELO_INTERVALS[Rank.STONE].max - lastPlayerElo + 1, // Assuming there could be players below 0 elo
    RANK_ELO_INTERVALS[Rank.BRONZE].size(),
    RANK_ELO_INTERVALS[Rank.SILVER].size(),
    RANK_ELO_INTERVALS[Rank.GOLD].size(),
    RANK_ELO_INTERVALS[Rank.PLATINUM].size(),
    RANK_ELO_INTERVALS[Rank.DIAMOND].size(),
    RANK_ELO_INTERVALS[Rank.MASTER].size(),
    lastAethereanElo - RANK_ELO_INTERVALS[Rank.GRANDMASTER].min + 1,
    firstPlayerElo - lastAethereanElo,
  ].map((eloIntervalSize) => (eloIntervalSize / totalEloRange) * totalWidth);
};

const formatBarData = (rawData: barDataItem[], barWidths: number[]) =>
  [...rawData].reverse().map((item, index) => ({
    ...item,
    barWidth: barWidths[index],
    topLabelComponent: () => <Text style={styles.topLabel}>{formatNumber(item.value ?? 0)}</Text>,
  }));

type Props = {
  width: number;
};

export const RankDistributionBarChart = ({ width }: Props) => {
  const {
    firstPlayerElo,
    lastPlayerElo,
    lastAethereanElo,
    rankDistribution,
    isLoading: isLoadingLeaderboard,
  } = useLeaderboardStats();

  if (isLoadingLeaderboard)
    return (
      <View style={{ height: width }}>
        <Spinner />
      </View>
    );

  const barWidths = getBarWidths(firstPlayerElo, lastPlayerElo, lastAethereanElo, width);

  return (
    <View testID="rank-distribution">
      <BarChart
        data={formatBarData(rankDistribution, barWidths)}
        disablePress
        disableScroll
        height={width}
        hideAxesAndRules
        spacing={0}
        width={width}
        xAxisLabelsHeight={0}
        yAxisLabelWidth={0}
      />
    </View>
  );
};

const styles = StyleSheet.create((theme) => ({
  topLabel: {
    fontFamily: theme.font.secondary.black,
    fontSize: 10,
    color: theme.color.black,
  },
}));
