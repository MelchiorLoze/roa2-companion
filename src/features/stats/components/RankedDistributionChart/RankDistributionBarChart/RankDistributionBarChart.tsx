import { Text, View } from 'react-native';
import { BarChart, type barDataItem } from 'react-native-gifted-charts';
import { StyleSheet, useUnistyles } from 'react-native-unistyles';

import { Spinner } from '@/components/Spinner/Spinner';

import { useLeaderboardStats } from '../../../hooks/business/useLeaderboardStats/useLeaderboardStats';
import { useRankDistribution } from '../../../hooks/business/useRankDistribution/useRankDistribution';
import { Rank, RANK_ELO_INTERVALS } from '../../../types/rank';

const formatNumber = (value: number) => {
  if (value >= 1_000_000) return `${(value / 1_000_000).toFixed(1)}M`;
  if (value >= 1_000) return `${(value / 1_000).toFixed(1)}K`;
  return value.toString();
};

// Calculate the widths of the chart bars to be proportional to the elo intervals
const getBarWidths = (firstPlayerElo: number, lastPlayerElo: number, lastAethereanElo: number, totalWidth: number) => {
  const totalEloRange = firstPlayerElo - lastPlayerElo + 1;
  return [
    RANK_ELO_INTERVALS[Rank.STONE].max - lastPlayerElo + 1, // Assuming there could be players below 0 elo
    RANK_ELO_INTERVALS[Rank.BRONZE].size(),
    RANK_ELO_INTERVALS[Rank.SILVER].size(),
    RANK_ELO_INTERVALS[Rank.GOLD].size(),
    RANK_ELO_INTERVALS[Rank.PLATINUM].size(),
    RANK_ELO_INTERVALS[Rank.DIAMOND].size(),
    RANK_ELO_INTERVALS[Rank.MASTER].size(),
    lastAethereanElo - RANK_ELO_INTERVALS[Rank.GRANDMASTER].min + 1,
    firstPlayerElo - lastAethereanElo,
  ].map((eloIntervalSize) => (eloIntervalSize / totalEloRange) * totalWidth);
};

const formatBarData = (rawData: Record<Rank, number>, barWidths: number[], theme: Theme) =>
  Object.entries(rawData).map<barDataItem>(([rank, count], index) => ({
    value: count,
    frontColor: theme.color[rank as Rank],
    barWidth: barWidths[index],
    topLabelComponent: () => <Text style={styles.topLabel}>{formatNumber(count)}</Text>,
  }));

type Props = {
  width: number;
};

export const RankDistributionBarChart = ({ width }: Readonly<Props>) => {
  const { firstPlayerElo, lastPlayerElo, lastAethereanElo } = useLeaderboardStats();
  const { rankDistribution, isLoading } = useRankDistribution();
  const { theme } = useUnistyles();

  if (isLoading)
    return (
      <View style={{ height: width }}>
        <Spinner />
      </View>
    );

  const barWidths = getBarWidths(firstPlayerElo, lastPlayerElo, lastAethereanElo, width);

  return (
    <View testID="rank-distribution">
      <BarChart
        data={formatBarData(rankDistribution, barWidths, theme)}
        disablePress
        disableScroll
        height={width}
        hideAxesAndRules
        spacing={0}
        width={width}
        xAxisLabelsHeight={0}
        yAxisLabelWidth={0}
      />
    </View>
  );
};

const styles = StyleSheet.create((theme) => ({
  topLabel: {
    fontFamily: theme.font.secondary.black,
    fontSize: 10,
    color: theme.color.black,
  },
}));
