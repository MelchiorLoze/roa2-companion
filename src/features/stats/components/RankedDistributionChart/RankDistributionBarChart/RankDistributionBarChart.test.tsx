import { render, screen } from '@testing-library/react-native';

import { useLeaderboardStats } from '../../../hooks/business/useLeaderboardStats/useLeaderboardStats';
import { testLeaderboardEntries } from '../../../test-helpers/testLeaderboardEntries';
import { RankDistributionBarChart } from './RankDistributionBarChart';

jest.mock('../../../hooks/business/useLeaderboardStats/useLeaderboardStats');
const useLeaderboardStatsMock = jest.mocked(useLeaderboardStats);
const defaultLeaderboardStatsState = {
  firstPlayerElo: 2162,
  lastPlayerElo: -100,
  lastAethereanElo: 1837,
  leaderboardEntries: testLeaderboardEntries,
  isLoading: false,
};

const renderComponent = () => {
  const result = render(<RankDistributionBarChart width={300} />);

  expect(useLeaderboardStatsMock).toHaveBeenCalledTimes(2);

  return result;
};

describe('RankDistributionBarChart', () => {
  beforeEach(() => {
    useLeaderboardStatsMock.mockReturnValue(defaultLeaderboardStatsState);
  });

  it('renders loading state when leaderboard stats are loading', () => {
    useLeaderboardStatsMock.mockReturnValue({
      ...defaultLeaderboardStatsState,
      isLoading: true,
    });

    renderComponent();

    screen.getByTestId('spinner');
    expect(screen.queryByTestId('rank-distribution')).toBeNull();
  });

  it('renders the chart when leaderboard stats are available', () => {
    const tree = renderComponent();

    screen.getByTestId('rank-distribution');
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
