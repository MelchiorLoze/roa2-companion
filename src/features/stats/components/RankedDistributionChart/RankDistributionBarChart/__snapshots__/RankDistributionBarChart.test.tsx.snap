// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RankDistributionBarChart renders the chart when leaderboard stats are available 1`] = `
<View
  testID="rank-distribution"
>
  <
    data={
      [
        {
          "barWidth": 79.54043305346885,
          "frontColor": "#7F7A66",
          "topLabelComponent": [Function],
          "value": 3,
        },
        {
          "barWidth": 26.513477684489615,
          "frontColor": "#B87A2E",
          "topLabelComponent": [Function],
          "value": 2,
        },
        {
          "barWidth": 26.513477684489615,
          "frontColor": "#C8C8D5",
          "topLabelComponent": [Function],
          "value": 2,
        },
        {
          "barWidth": 26.513477684489615,
          "frontColor": "#FFC64E",
          "topLabelComponent": [Function],
          "value": 2,
        },
        {
          "barWidth": 26.513477684489615,
          "frontColor": "#C3C4FF",
          "topLabelComponent": [Function],
          "value": 3,
        },
        {
          "barWidth": 26.513477684489615,
          "frontColor": "#95DFEC",
          "topLabelComponent": [Function],
          "value": 1,
        },
        {
          "barWidth": 26.513477684489615,
          "frontColor": "#9CE7AC",
          "topLabelComponent": [Function],
          "value": 3,
        },
        {
          "barWidth": 18.294299602297833,
          "frontColor": "#FC7575",
          "topLabelComponent": [Function],
          "value": 1,
        },
        {
          "barWidth": 43.08440123729562,
          "frontColor": "#BD7CC5",
          "topLabelComponent": [Function],
          "value": 4,
        },
      ]
    }
    disablePress={true}
    disableScroll={true}
    height={300}
    hideAxesAndRules={true}
    spacing={0}
    width={300}
    xAxisLabelsHeight={0}
    yAxisLabelWidth={0}
  />
</View>
`;
